DeepSpeech.py

```diff
@@ -396,29 +395,89 @@ def log_grads_and_vars(grads_and_vars):
-    try:
-        checkpoint = tf.train.get_checkpoint_state(FLAGS.checkpoint_dir, checkpoint_filename)
-        if not checkpoint:
-            return False
-        checkpoint_path = checkpoint.model_checkpoint_path
-        saver.restore(session, checkpoint_path)
-        if load_step:
-            restored_step = session.run(tfv1.train.get_global_step())
-            if log_success:
-                log_info('Restored variables from %s checkpoint at %s, step %d' %
-                         (caption, checkpoint_path, restored_step))
-        elif log_success:
-            log_info('Restored variables from %s checkpoint at %s' % (caption, checkpoint_path))
-        return True
-    except tf.errors.InvalidArgumentError as e:
-        log_error(str(e))
-        log_error('The checkpoint in {0} does not match the shapes of the model.'
-                  ' Did you change alphabet.txt or the --n_hidden parameter'
-                  ' between train runs using the same checkpoint dir? Try moving'
-                  ' or removing the contents of {0}.'.format(checkpoint_path))
+
+def load_model(session, saver, checkpoint_filename, caption, drop_source_layers, load_step=True, log_success=True, use_cudnn):
+
+'''
+load+list = [put all vars here first]
+init_list=[]
+
+if use_cudnn:
+     move all adam vars to init_list
+if drop_souce_layers>0:
+     move all drop layers to init_list
+if init:
+   override! initialize all vars
+'''
+
+    if FLAGS.load == 'init':
+        session.run(tfv1.global_variables_initializer())
+        return
+    
+    ckpt = tfv1.train.load_checkpoint(checkpoint_filename)
+    load_vars = set(tfv1.global_variables())
+    init_vars = set()
+
+    if use_cudnn:
+        # Load compatible variables from checkpoint    
+        for v in load_vars:
+            try:
+                ckpt.get_tensor(v.op.name)
+            except tf.errors.NotFoundError:
+                load_vars -= v
+                init_vars.add(v)
+
+        # Check that the only missing variables (i.e. those to be initialised)
+        # are the Adam moment tensors
+        if any('Adam' not in v.op.name for v in init_vars):
+            log_error('Tried to load a CuDNN RNN checkpoint but there were '
+                      'more missing variables than just the Adam moment '
+                      'tensors.')
+            sys.exit(1)
+
+    if drop_source_layers>0:
+        '''
+        The transfer learning approach here need us to supply the layers which we
+        want to exclude from the source model.
+        Say we want to exclude all layers except for the first one, we can use this:
+        drop_source_layers=['2', '3', 'lstm', '5', '6']
+        If we want to use all layers from the source model except the last one, we use this:
+        drop_source_layers=['6']
+        '''
+        # FIXME -- if someone tries to drop 6+ layers, this will return an empty list
```

Comment:
nit: `-1 * min(6, int(drop_source_layers))` in place of `-int(drop_source_layers)` ?

Reply:
`-1 * min(5, int(drop_source_layers))` is better, with a warning message. If the user will try to drop `6` layers, then we will have a fresh init

Reply:
> `-1 * min(5, int(drop_source_layers))` is better, with a warning message. If the user will try to drop `6` layers, then we will have a fresh init

Let's fold the min here and document that properly in the help message, I guess we don't need more.

Reply:
fixed